src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContextHelper.cs

```diff
@@ -0,0 +1,167 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeFixes
+{
+    internal static partial class FixAllContextHelper
+    {
+        public static async Task<ImmutableDictionary<Document, ImmutableArray<Diagnostic>>> GetDocumentDiagnosticsToFixAsync(
+            FixAllContext fixAllContext)
+        {
+            var cancellationToken = fixAllContext.CancellationToken;
+
+            var allDiagnostics = ImmutableArray<Diagnostic>.Empty;
+
+            var document = fixAllContext.Document;
+            var project = fixAllContext.Project;
+
+            var progressTracker = fixAllContext.GetProgressTracker();
+
+            switch (fixAllContext.Scope)
+            {
+                case FixAllScope.Document:
+                    if (document != null && !await document.IsGeneratedCodeAsync(cancellationToken).ConfigureAwait(false))
+                    {
+                        var documentDiagnostics = await fixAllContext.GetDocumentDiagnosticsAsync(document).ConfigureAwait(false);
+                        return ImmutableDictionary<Document, ImmutableArray<Diagnostic>>.Empty.SetItem(document, documentDiagnostics);
+                    }
+
+                    break;
+
+                case FixAllScope.ContainingMember or FixAllScope.ContainingType:
+                    if (document != null && !await document.IsGeneratedCodeAsync(cancellationToken).ConfigureAwait(false))
+                    {
+                        // For the 'ContainingMember' and 'ContainingType' scopes, we might be invoked in a couple of different ways:
+                        //  1. Using the fully populated 'FixAllContext', which has already computed, non-empty 'fixAllSpans' for
+                        //     the member or type declaration to fix in the given 'document'. For this scenario, we return the
+                        //     diagnostics for the given fixAllSpans in the document for this scenario.
+                        //  2. Using the original 'FixAllContext' which has the 'triggerSpan' and 'document' for the original code fix, but
+                        //     the 'fixAllSpans' have not yet been computed. For this scenario, we use the 'IFixAllSpanMappingService' to
+                        //     map the triggerSpan and fixAllScope to documents and spans to fix for the containing member or
+                        //     type declaration (and its partials), and then return the diagnostics for each of the documents and fixAllSpans.
```

Comment:
this is interesting and thanks for the info.  but could you clarify why some paths have a `fully populated 'FixAllContext', which has already computed, non-empty 'fixAllSpans"`, while some don't?  It's genuinely not clear to me how that would happen.

Reply:
Thanks! So, there seemed to be couple of different code paths in to reach this helper and at some place the FixAllSpans were lazily computed and at some they were precomputed and passed around. I have cleaned this up by completely removing `FixAllState.FixAllSpans` property and instead we now always lazily compute these spans right when they are required to compute diagnostics to fix.